File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/dialog/editMeals/EditMealsDialog.java
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog.editMeals;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.TrackerApplication;
import de.karzek.diettracker.presentation.common.BaseDialog;
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog.editMeals.adapter.EditMealsListAdapter;
import de.karzek.diettracker.presentation.model.MealDisplayModel;

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
public class EditMealsDialog extends BaseDialog implements EditMealsDialogContract.View {

    public static final String EXTRA_SELECTED_MEALS = "EXTRA_SELECTED_MEALS";

    @Inject
    EditMealsDialogContract.Presenter presenter;

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    @BindView(R.id.loading_view)
    FrameLayout loadingView;

    @BindView(R.id.dialog_action_dismiss)
    Button dismiss;
    @BindView(R.id.dialog_action_save)
    Button save;

    private View view;

    private ArrayList<MealDisplayModel> mealList = new ArrayList<>();

    private SaveMealsSelectionDialogListener listener;

    @Override
    protected void setupDialogComponent() {
        TrackerApplication.get(getContext()).createEditMealsDialogComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.dialog_edit_checkbox_list, container, false);
        ButterKnife.bind(this, view);

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            mealList = bundle.getParcelableArrayList(EXTRA_SELECTED_MEALS);
        }
        presenter.setSelectedMealList(mealList);

        setupRecyclerView();

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                listener = null;
            }
        });

        save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                listener.updateMeals(presenter.getSelectedMeals());
                listener = null;
            }
        });

        presenter.setView(this);
        presenter.start();

        return view;
    }

    private void setupRecyclerView() {
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new EditMealsListAdapter(presenter));
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            listener = (SaveMealsSelectionDialogListener) getActivity();
        } catch (ClassCastException e) {
            throw new ClassCastException("fragment must implement SaveMealsSelectionDialogListener");
        }
    }

    @Override
    public void updateRecyclerView(ArrayList<MealDisplayModel> meals, HashMap<Integer, Boolean> mealStatus) {
        ((EditMealsListAdapter) recyclerView.getAdapter()).setMeals(meals, mealStatus);
    }

    @Override
    public void showLoading() {
        loadingView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        loadingView.setVisibility(View.GONE);
    }

    @OnClick(R.id.reset_selection)
    void onResetSelectionClicked() {
        presenter.onResetSelectionClicked();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.finish();
        TrackerApplication.get(getContext()).releaseEditMealsDialogComponent();
    }

    public interface SaveMealsSelectionDialogListener {
        void updateMeals(ArrayList<MealDisplayModel> selectedMeals);
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/dialog/editMeals/EditMealsDialog.kt
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog.editMeals

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.TrackerApplication
import de.karzek.diettracker.presentation.common.BaseDialog
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog.editMeals.adapter.EditMealsListAdapter
import de.karzek.diettracker.presentation.model.MealDisplayModel
import kotlinx.android.synthetic.main.dialog_edit_checkbox_list.*
import kotlinx.android.synthetic.main.snippet_loading_view.*
import java.util.*
import javax.inject.Inject

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
class EditMealsDialog : BaseDialog(), EditMealsDialogContract.View {

    @Inject
    lateinit var presenter: EditMealsDialogContract.Presenter

    private var dialogView: View? = null

    private var mealList: ArrayList<MealDisplayModel>? = ArrayList()

    private var listener: SaveMealsSelectionDialogListener? = null

    override fun setupDialogComponent() {
        TrackerApplication.get(context!!).createEditMealsDialogComponent().inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialogView = inflater.inflate(R.layout.dialog_edit_checkbox_list, container, false)

        val bundle = this.arguments
        if (bundle != null) {
            mealList = bundle.getParcelableArrayList(EXTRA_SELECTED_MEALS)
        }
        presenter.setSelectedMealList(mealList)

        setupRecyclerView()

        dialog_action_dismiss.setOnClickListener {
            dismiss()
            listener = null
        }

        dialog_action_save.setOnClickListener {
            dismiss()
            listener!!.updateMeals(presenter.selectedMeals)
            listener = null
        }

        reset_selection.setOnClickListener { presenter.onResetSelectionClicked() }

        presenter.setView(this)
        presenter.start()

        return dialogView
    }

    private fun setupRecyclerView() {
        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(context)
        recycler_view.adapter = EditMealsListAdapter(presenter)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            listener = activity as SaveMealsSelectionDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("fragment must implement SaveMealsSelectionDialogListener")
        }

    }

    override fun updateRecyclerView(meals: ArrayList<MealDisplayModel>, mealStatus: HashMap<Int, Boolean>) {
        (recycler_view.adapter as EditMealsListAdapter).setMeals(meals, mealStatus)
    }

    override fun showLoading() {
        loading_view.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loading_view.visibility = View.GONE
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.finish()
        TrackerApplication.get(context!!).releaseEditMealsDialogComponent()
    }

    interface SaveMealsSelectionDialogListener {
        fun updateMeals(selectedMeals: ArrayList<MealDisplayModel>)
    }

    companion object {

        const val EXTRA_SELECTED_MEALS = "EXTRA_SELECTED_MEALS"
    }
}
